src/System.Net.Security/src/System/Net/_SslSessionsCache.cs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+
+namespace System.Net.Security
+{
+    // Implements SSL session caching mechanism based on a static table of SSL credentials.
+    internal static class SslSessionsCache
+    {
+        private const int c_CheckExpiredModulo = 32;
+        private static Hashtable s_CachedCreds = new Hashtable(32);
+
+        //
+        // Uses cryptographically strong certificate thumbprint comparison.
+        //
+        private struct SslCredKey
+        {
+            private byte[] _CertThumbPrint;
+            private int _AllowedProtocols;
+            private EncryptionPolicy _EncryptionPolicy;
+            private readonly int _HashCode;
+
+            //
+            // SECURITY: X509Certificate.GetCertHash() is virtual hence before going here,
+            //           the caller of this ctor has to ensure that a user cert object was inspected and
+            //           optionally cloned.
+            //
+            internal SslCredKey(byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
```

Comment:
While it's Hard(tm) for two legitimate certificates to end up with the same thumbprint, it's not guaranteed unique.  If you think the domain is constricted enough to make you feel safe, that's fine; otherwise some additional material should be used as a tiebreaker.